tasty/src/dotty/tools/tasty/TastyHeaderUnpickler.scala

```diff
@@ -0,0 +1,136 @@
+      throw new UnpickleException(signature + backIncompatAddendum + myAddendum)
+    }
+    else {
+      val fileMinor = readNat()
+      val fileExperimental = readNat()
+      val toolingVersion = {
+        val length = readNat()
+        val start = currentAddr
+        val end = start + length
+        goto(end)
+        new String(bytes, start.index, length) // NOTE: new string object here (should probably cache as it leaks)
+      }
+
+      val validVersion = ( // inlined from `dotty.tools.tasty.TastyVersionFormatTest.TastyVersion.<:<`
+        fileMajor == MajorVersion && (
+          if (fileExperimental == ExperimentalVersion) {
+            if (ExperimentalVersion == 0) {
+              fileMinor <= MinorVersion
+            }
+            else {
+              fileMinor == MinorVersion
+            }
+          }
+          else {
+            fileExperimental == 0 && fileMinor < MinorVersion
+          }
+        )
+      )
+      check(validVersion, {
+        val foundMinor = showMinorVersion(fileMinor, fileExperimental)
+        val producedByAddendum = s"\nThe TASTy file was produced by $toolingVersion.$myAddendum"
+        val signature = signatureString(expectedMinor, fileMajor, foundMinor)
+        if (fileExperimental != 0)
+          signature + unstableAddendum + producedByAddendum
+        else if (fileMajor < MajorVersion)
+          signature + backIncompatAddendum + producedByAddendum
+        else
+          signature + forwardIncompatAddendum + producedByAddendum
+      }
+      )
+      val uuid = new UUID(readUncompressedLong(), readUncompressedLong())
+      new TastyHeader(uuid, fileMajor, fileMinor, fileExperimental, toolingVersion) {}
+    }
+  }
+
+  def isAtEnd: Boolean = reader.isAtEnd
+
+  private def check(cond: Boolean, msg: => String): Unit = {
+    if (!cond) throw new UnpickleException(msg)
+  }
+}
+
+object TastyHeaderUnpickler {
+
+  private def showMinorVersion(min: Int, exp: Int) = {
+    val expStr = if (exp == 0) "" else s" [unstable release: $exp]"
+    s"$min$expStr"
+  }
+
+  private def expectedMinor = showMinorVersion(MinorVersion, ExperimentalVersion)
+
+  private def myAddendum = {
+    if (ExperimentalVersion > 0)
+      "\nNote that your tooling is currently using an unstable TASTy version."
+    else
+      ""
+  }
+
+  private def signatureString(minorVersion: String, fileMajor: Int, fileMinor: String) = (
+    s"""TASTy signature has wrong version.
+      | expected: {majorVersion: $MajorVersion, minorVersion: $minorVersion}
+      | found   : {majorVersion: $fileMajor, minorVersion: $fileMinor}
+      |
+      |""".stripMargin
+  )
+
+  private def unstableAddendum =
+    """This TASTy file was produced by an unstable release.
+      |To read this TASTy file, your tooling must be at the same version.""".stripMargin
+
+  private def backIncompatAddendum =
+    """This TASTy file was produced by a backwards incompatible release.
```

Comment:
It's this version which backward incompatible with the old version, so that error message is weird. Consider:
```suggestion
    """This TASTy file was produced by an earlier release that is not supported anymore.
```

Reply:
done